code-gen: reusable main.go logic, preparation for kube-gen

sttts opened the pull request:
Split main.go files into plumbing and calls to NewDefaults, AddFlags and Validate. This will allow us to create kube-gen without duplicating much code, at least no generator logic (with the exception of a little loop in client-gen adding the group path to the input dirs).

sttts committed:
client-gen: clarify clientset-api-path

sttts committed:
client-gen: separate input-base logic from CustomArgs

sttts committed:
code-gen: uniform reusable main.go logic

Split main.go into plumbing and calls to NewDefaults, AddFlags and Validate.

sttts committed:
client-gen: use --output-package instead of --clientset-path

sttts committed:
Update bazel

antoineco commented on staging/src/k8s.io/code-generator/cmd/client-gen/args/args.go:
```diff
@@ -53,13 +56,70 @@ type CustomArgs struct {
-func (ca *CustomArgs) AddFlags(fs *pflag.FlagSet) {
-	pflag.Var(NewGVPackagesValue(&ca.GroupVersionToInputPath, &ca.Groups, nil), "input", "group/versions that client-gen will generate clients for. At most one version per group is allowed. Specified in the format \"group1/version1,group2/version2...\".")
+func NewDefaults() (*args.GeneratorArgs, *CustomArgs) {
+	genericArgs := args.Default().WithoutDefaultFlagParsing()
```
Doesn't `WithoutDefaultFlagParsing()` prevent passing `glog` flags?

sttts replied:
What should it? We add the Go flags (including glog) to the flag set.

antoineco replied:
`--logtostderr` for example. If I make it the default we may want to disable it. The `-v` flag is another example. AFAIK only the standard `flags` package supports these flags, and pflags requires you to explicitly copy the `flags.CommandLine` flagset, hence the need for the default flagset.

antoineco replied:
OK scratch that, there's a call to `AddGoFlagSet()` in `main.go` (y)

sttts replied:
:)

pwittrock commented:
/assign @droot

pwittrock commented:
/assign @jennybuckley

sttts commented:
/retest

pwittrock commented on hack/update-codegen.sh:
```diff
@@ -72,7 +72,7 @@ INTERNAL_DIRS_CSV=$(IFS=',';echo "${INTERNAL_DIRS[*]// /,}";IFS=$)
 # update- and verify- scripts.
 ${clientgen} --input-base="k8s.io/kubernetes/pkg/apis" --input="${INTERNAL_DIRS_CSV}" "$@"
-${clientgen} --output-base "${KUBE_ROOT}/vendor" --clientset-path="k8s.io/client-go" --clientset-name="kubernetes" --input-base="k8s.io/kubernetes/vendor/k8s.io/api" --input="${GV_DIRS_CSV}" "$@"
+${clientgen} --output-base "${KUBE_ROOT}/vendor" --output-package="k8s.io/client-go" --clientset-name="kubernetes" --input-base="k8s.io/kubernetes/vendor/k8s.io/api" --input="${GV_DIRS_CSV}" "$@"
```
this is a better name :)

pwittrock commented on staging/src/k8s.io/code-generator/cmd/conversion-gen/args/args.go:
```diff
@@ -0,0 +1,81 @@
+// CustomArgs is used by the gengo framework to pass args specific to this generator.
+type CustomArgs struct {
+	// Base peer dirs which nearly everybody will use, i.e. outside of Kubernetes core.
+	BasePeerDirs []string
```
Can you clarify what a "PeerDir" is as part of the comment.  It may not be obvious to everyone.

sttts replied:
done

pwittrock commented on staging/src/k8s.io/code-generator/cmd/conversion-gen/args/args.go:
```diff
@@ -0,0 +1,81 @@
+	// of these operations. The unsafe operation is a direct pointer assignment via unsafe
+	// (within the allowed uses of unsafe) and is equivalent to a proposed Golang change to
+	// allow structs that are identical to be assigned to each other.
+	SkipUnsafe bool
```
What should the default value be, and why?

sttts replied:
looks like we disable it in the generator at the moment, but enable it in kube. In theory, every (wrong) unsafe conversion can cause panics, i.e. it's not type checked at all. So we depend 100% on the generator to apply them correctly. I tend to turn them on, but do not know whether there are corner cases where this harms.

pwittrock commented on staging/src/k8s.io/code-generator/pkg/util/build.go:
```diff
@@ -0,0 +1,49 @@
+// CurrentPackage returns the go package of the current directory, or "" if it cannot
+// be derived from the GOPATH.
+func CurrentPackage() string {
+	for _, root := range gobuild.Default.SrcDirs() {
```
This is cool

pwittrock commented on staging/src/k8s.io/code-generator/pkg/util/build.go:
```diff
@@ -0,0 +1,49 @@
+}
+
+func hasSubdir(root, dir string) (rel string, ok bool) {
+	const sep = string(filepath.Separator)
```
Can you add comments for each of these blocks, it isn't obvious what this is supposed to do

sttts replied:
done

sttts committed:
conversion-gen: add godocs for peer dirs

sttts commented:
@pwittrock addressed your comments. ptal.

If you start working on kubegen, just base it on this PR for the moment until master is open again and we can merge it.

sttts commented:
@pwittrock let me know if you see any other blockers for your work.

pwittrock commented:
@sttts thanks

pwittrock commented:
/retest

pwittrock commented:
@sttts The PR looks good to me, but I am definitely not an expert in this area.

pwittrock commented:
Would you also add a release note if needed.

sttts commented:
> Would you also add a release note if needed.

It does not change the user+admin experience. Usually we do not add release-notes then.

We really need a `developer-release-note` field in PRs and matching labels. Wdyt?

pwittrock commented:
> We really need a developer-release-note field in PRs and matching labels. Wdyt?

@sttts 

Probably would be useful.  Why don't you add release-note-none for this PR and we can merge it.  Has master been opened?

pwittrock commented:
/lgtm

pwittrock commented:
/approve

sttts commented:
/approve no-issue

sttts commented on staging/BUILD:
```diff
@@ -210,6 +210,7 @@ filegroup(
         "//staging/src/k8s.io/code-generator/cmd/lister-gen:all-srcs",
         "//staging/src/k8s.io/code-generator/cmd/openapi-gen:all-srcs",
         "//staging/src/k8s.io/code-generator/cmd/set-gen:all-srcs",
+        "//staging/src/k8s.io/code-generator/pkg/util:all-srcs",
```
@mikedanese this is ugly that every change in some deeply nested staging repo requires approval on staging/ level.

sttts commented:
Overriding staging/ approval as it is only about BUILD (why does this need approvals at all?)
